pkg/api-server/dataplane_layout_endpoint.go

```diff
@@ -0,0 +1,186 @@
+package api_server
+
+import (
+	"context"
+	"net/http"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/emicklei/go-restful/v3"
+
+	"github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/api/openapi/types"
+	api_common "github.com/kumahq/kuma/api/openapi/types/common"
+	"github.com/kumahq/kuma/pkg/core/kri"
+	"github.com/kumahq/kuma/pkg/core/resources/access"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/core"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	rest_errors "github.com/kumahq/kuma/pkg/core/rest/errors"
+	"github.com/kumahq/kuma/pkg/core/user"
+	util_slices "github.com/kumahq/kuma/pkg/util/slices"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+)
+
+type dataplaneLayoutEndpoint struct {
+	resManager         manager.ResourceManager
+	meshContextBuilder xds_context.MeshContextBuilder
+	resourceAccess     access.ResourceAccess
+}
+
+func newDataplaneLayoutEndpoint(resManager manager.ResourceManager, meshContextBuilder xds_context.MeshContextBuilder, resourceAccess access.ResourceAccess) *dataplaneLayoutEndpoint {
+	return &dataplaneLayoutEndpoint{
+		resManager:         resManager,
+		meshContextBuilder: meshContextBuilder,
+		resourceAccess:     resourceAccess,
+	}
+}
+
+func (dle *dataplaneLayoutEndpoint) addEndpoint(ws *restful.WebService) {
+	ws.Route(
+		ws.GET("/meshes/{mesh}/dataplanes/{name}/_layout").
+			To(dle.getLayout).
+			Doc("Get Dataplane Layout").
+			Returns(http.StatusOK, "OK", nil),
+	)
+}
+
+func (dle *dataplaneLayoutEndpoint) getLayout(request *restful.Request, response *restful.Response) {
+	meshName := request.PathParameter("mesh")
+	dataplaneName := request.PathParameter("name")
+
+	if err := dle.resourceAccess.ValidateGet(
+		request.Request.Context(),
+		core_model.ResourceKey{Mesh: meshName, Name: dataplaneName},
+		core_mesh.DataplaneResourceTypeDescriptor,
+		user.FromCtx(request.Request.Context()),
+	); err != nil {
+		rest_errors.HandleError(request.Request.Context(), response, err, "Access Denied")
+		return
+	}
+
+	mesh, err := dle.getMeshResource(request.Request.Context(), meshName)
+	if err != nil {
+		rest_errors.HandleError(request.Request.Context(), response, err, "Failed to retrieve Mesh")
+		return
+	}
+
+	if mesh.Spec.GetMeshServices().GetMode() == v1alpha1.Mesh_MeshServices_Disabled {
+		response.WriteHeader(http.StatusBadRequest)
```

Comment:
I think this should be:

```
rest_errors.HandleError(..., response, errors.NewBadRequest("can't use _layout endpoint without meshService"))
```

Otherwise it's likely not very straightforward for users to understand why this isn't valid and it's also not compliant with the AIP: https://kong-aip.netlify.app/aip/193/